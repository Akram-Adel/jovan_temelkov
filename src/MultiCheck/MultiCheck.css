.MultiCheck .flex-row {
  display: flex;
  flex-direction: row;
}

.MultiCheck .column {
  padding-right: 20px;
}
