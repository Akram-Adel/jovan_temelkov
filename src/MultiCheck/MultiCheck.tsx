import './MultiCheck.css';

import React from 'react';

import CheckOption from './components/CheckOption';
import ToColumns from './components/ToColumns';

export type Option = {
  label: string;
  value: string;
};

/**
 * Notice:
 * 1. There should be a special `Select All` option with checkbox to control all passing options
 * 2. If columns > 1, the options should be placed from top to bottom in each column
 *
 * @param {string} label - the label text of this component
 * @param {Option[]} options - options
 * @param {string[]} values - default checked option values
 * @param {number} columns - default value is 1
 * @param {Function} onChange - when checked options are changed,
 *                             they should be passed to outside
 */
type Props = {
  label?: string;
  options: Option[];
  columns?: number;
  values?: string[];
  onChange?: (values: string[]) => void;
};

const MultiCheck: React.FunctionComponent<Props> = (props): JSX.Element => {
  function onOptionCheck(optionValue: Option['value']) {
    return function onChange(checked: boolean) {
      props.onChange?.(
        !checked
          ? (props.values ?? []).filter((value) => value !== optionValue)
          : [...(props.values ?? []), optionValue],
      );
    };
  }

  function toggleSelectAll(selectAll: boolean) {
    props.onChange?.(selectAll ? props.options.map((option) => option.value) : []);
  }

  return (
    <div className="MultiCheck">
      {props.label ? <h2>{props.label}</h2> : null}

      <ToColumns columns={props.columns ?? 1}>
        <CheckOption
          label="Select All"
          value="select-all"
          checked={props.options.length === props.values?.length}
          onChange={toggleSelectAll}
        />

        {props.options.map(({label, value}) => (
          <CheckOption
            key={value}
            {...{label, value}}
            checked={props.values?.includes(value)}
            onChange={onOptionCheck(value)}
          />
        ))}
      </ToColumns>
    </div>
  );
};

export default MultiCheck;
