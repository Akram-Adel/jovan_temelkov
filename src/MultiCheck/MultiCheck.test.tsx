import '@testing-library/jest-dom';

import {fireEvent, getAllByRole, render} from '@testing-library/react';
import React from 'react';

import MultiCheck, {Option} from './MultiCheck';

const options: Option[] = [
  {label: 'aaa', value: '111'},
  {label: 'bbb', value: '222'},
  {label: 'ccc', value: '333'},
  {label: 'ddd', value: '444'},
  {label: 'eee', value: '555'},
  {label: 'fff', value: '666'},
  {label: 'ggg', value: '777'},
  {label: 'hhh', value: '888'},
  {label: 'iii', value: '999'},
];

describe('MultiCheck', () => {
  describe('initialize', () => {
    it('renders correctly without the optional props', () => {
      expect(render(<MultiCheck options={options} />).container).toMatchSnapshot();
    });

    it('renders the label if label provided', () => {
      expect(render(<MultiCheck label="my-multi-check" options={options} />).container).toMatchSnapshot();
    });

    it('renders different column configurations correctly', () => {
      expect(render(<MultiCheck options={options} columns={1} />).container).toMatchSnapshot();
      expect(render(<MultiCheck options={options} columns={2} />).container).toMatchSnapshot();
      expect(render(<MultiCheck options={options} columns={3} />).container).toMatchSnapshot();
    });
  });

  describe('base logic', () => {
    it('checks provided values correctly', () => {
      const numOfCheckedOptions = 2;

      const {container} = render(
        <MultiCheck
          options={options}
          values={options.slice(0, numOfCheckedOptions).map((option) => option.value)}
        />,
      );

      getAllByRole(container, 'checkbox').forEach((element, index) => {
        if (index === 0 || index > numOfCheckedOptions) {
          expect(element).not.toBeChecked();
        } else {
          expect(element).toBeChecked();
        }
      });
    });

    it('returns all checked values correctly when they change', () => {
      const onChange = jest.fn();
      const {container} = render(<MultiCheck options={options} onChange={onChange} />);

      fireEvent.click(getAllByRole(container, 'checkbox')[1]);

      expect(onChange).toHaveBeenCalledWith([options[0].value]);
    });
  });

  describe('select-all logic', () => {
    it('uncheckes select-all by default if any other option is unchecked', () => {
      const {container} = render(
        <MultiCheck options={options} values={options.slice(1).map((option) => option.value)} />,
      );

      expect(getAllByRole(container, 'checkbox')[0]).not.toBeChecked();
    });

    it('returns all options when select-all is checked', () => {
      const onChange = jest.fn();
      const {container} = render(<MultiCheck options={options} onChange={onChange} />);

      fireEvent.click(getAllByRole(container, 'checkbox')[0]);

      expect(onChange).toHaveBeenCalledWith(options.map((option) => option.value));
    });

    it('checkes select-all by default if all other options are checked', () => {
      const {container} = render(
        <MultiCheck options={options} values={options.map((option) => option.value)} />,
      );

      expect(getAllByRole(container, 'checkbox')[0]).toBeChecked();
    });

    it('returns no options when select-all is unchecked', () => {
      const onChange = jest.fn();
      const {container} = render(
        <MultiCheck options={options} onChange={onChange} values={options.map((option) => option.value)} />,
      );

      fireEvent.click(getAllByRole(container, 'checkbox')[0]);

      expect(onChange).toHaveBeenCalledWith([]);
    });
  });
});
