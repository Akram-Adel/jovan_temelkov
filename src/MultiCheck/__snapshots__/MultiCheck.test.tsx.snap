// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiCheck initialize renders correctly without the optional props 1`] = `
<div>
  <div
    class="MultiCheck"
  >
    <div
      class="flex-row"
    >
      <div
        class="column"
      >
        <div>
          <input
            id="select-all"
            type="checkbox"
            value="select-all"
          />
          <label
            for="select-all"
          >
             
            Select All
          </label>
        </div>
        <div>
          <input
            id="111"
            type="checkbox"
            value="111"
          />
          <label
            for="111"
          >
             
            aaa
          </label>
        </div>
        <div>
          <input
            id="222"
            type="checkbox"
            value="222"
          />
          <label
            for="222"
          >
             
            bbb
          </label>
        </div>
        <div>
          <input
            id="333"
            type="checkbox"
            value="333"
          />
          <label
            for="333"
          >
             
            ccc
          </label>
        </div>
        <div>
          <input
            id="444"
            type="checkbox"
            value="444"
          />
          <label
            for="444"
          >
             
            ddd
          </label>
        </div>
        <div>
          <input
            id="555"
            type="checkbox"
            value="555"
          />
          <label
            for="555"
          >
             
            eee
          </label>
        </div>
        <div>
          <input
            id="666"
            type="checkbox"
            value="666"
          />
          <label
            for="666"
          >
             
            fff
          </label>
        </div>
        <div>
          <input
            id="777"
            type="checkbox"
            value="777"
          />
          <label
            for="777"
          >
             
            ggg
          </label>
        </div>
        <div>
          <input
            id="888"
            type="checkbox"
            value="888"
          />
          <label
            for="888"
          >
             
            hhh
          </label>
        </div>
        <div>
          <input
            id="999"
            type="checkbox"
            value="999"
          />
          <label
            for="999"
          >
             
            iii
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MultiCheck initialize renders different column configurations correctly 1`] = `
<div>
  <div
    class="MultiCheck"
  >
    <div
      class="flex-row"
    >
      <div
        class="column"
      >
        <div>
          <input
            id="select-all"
            type="checkbox"
            value="select-all"
          />
          <label
            for="select-all"
          >
             
            Select All
          </label>
        </div>
        <div>
          <input
            id="111"
            type="checkbox"
            value="111"
          />
          <label
            for="111"
          >
             
            aaa
          </label>
        </div>
        <div>
          <input
            id="222"
            type="checkbox"
            value="222"
          />
          <label
            for="222"
          >
             
            bbb
          </label>
        </div>
        <div>
          <input
            id="333"
            type="checkbox"
            value="333"
          />
          <label
            for="333"
          >
             
            ccc
          </label>
        </div>
        <div>
          <input
            id="444"
            type="checkbox"
            value="444"
          />
          <label
            for="444"
          >
             
            ddd
          </label>
        </div>
        <div>
          <input
            id="555"
            type="checkbox"
            value="555"
          />
          <label
            for="555"
          >
             
            eee
          </label>
        </div>
        <div>
          <input
            id="666"
            type="checkbox"
            value="666"
          />
          <label
            for="666"
          >
             
            fff
          </label>
        </div>
        <div>
          <input
            id="777"
            type="checkbox"
            value="777"
          />
          <label
            for="777"
          >
             
            ggg
          </label>
        </div>
        <div>
          <input
            id="888"
            type="checkbox"
            value="888"
          />
          <label
            for="888"
          >
             
            hhh
          </label>
        </div>
        <div>
          <input
            id="999"
            type="checkbox"
            value="999"
          />
          <label
            for="999"
          >
             
            iii
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MultiCheck initialize renders different column configurations correctly 2`] = `
<div>
  <div
    class="MultiCheck"
  >
    <div
      class="flex-row"
    >
      <div
        class="column"
      >
        <div>
          <input
            id="select-all"
            type="checkbox"
            value="select-all"
          />
          <label
            for="select-all"
          >
             
            Select All
          </label>
        </div>
        <div>
          <input
            id="111"
            type="checkbox"
            value="111"
          />
          <label
            for="111"
          >
             
            aaa
          </label>
        </div>
        <div>
          <input
            id="222"
            type="checkbox"
            value="222"
          />
          <label
            for="222"
          >
             
            bbb
          </label>
        </div>
        <div>
          <input
            id="333"
            type="checkbox"
            value="333"
          />
          <label
            for="333"
          >
             
            ccc
          </label>
        </div>
        <div>
          <input
            id="444"
            type="checkbox"
            value="444"
          />
          <label
            for="444"
          >
             
            ddd
          </label>
        </div>
      </div>
      <div
        class="column"
      >
        <div>
          <input
            id="555"
            type="checkbox"
            value="555"
          />
          <label
            for="555"
          >
             
            eee
          </label>
        </div>
        <div>
          <input
            id="666"
            type="checkbox"
            value="666"
          />
          <label
            for="666"
          >
             
            fff
          </label>
        </div>
        <div>
          <input
            id="777"
            type="checkbox"
            value="777"
          />
          <label
            for="777"
          >
             
            ggg
          </label>
        </div>
        <div>
          <input
            id="888"
            type="checkbox"
            value="888"
          />
          <label
            for="888"
          >
             
            hhh
          </label>
        </div>
        <div>
          <input
            id="999"
            type="checkbox"
            value="999"
          />
          <label
            for="999"
          >
             
            iii
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MultiCheck initialize renders different column configurations correctly 3`] = `
<div>
  <div
    class="MultiCheck"
  >
    <div
      class="flex-row"
    >
      <div
        class="column"
      >
        <div>
          <input
            id="select-all"
            type="checkbox"
            value="select-all"
          />
          <label
            for="select-all"
          >
             
            Select All
          </label>
        </div>
        <div>
          <input
            id="111"
            type="checkbox"
            value="111"
          />
          <label
            for="111"
          >
             
            aaa
          </label>
        </div>
        <div>
          <input
            id="222"
            type="checkbox"
            value="222"
          />
          <label
            for="222"
          >
             
            bbb
          </label>
        </div>
        <div>
          <input
            id="333"
            type="checkbox"
            value="333"
          />
          <label
            for="333"
          >
             
            ccc
          </label>
        </div>
      </div>
      <div
        class="column"
      >
        <div>
          <input
            id="444"
            type="checkbox"
            value="444"
          />
          <label
            for="444"
          >
             
            ddd
          </label>
        </div>
        <div>
          <input
            id="555"
            type="checkbox"
            value="555"
          />
          <label
            for="555"
          >
             
            eee
          </label>
        </div>
        <div>
          <input
            id="666"
            type="checkbox"
            value="666"
          />
          <label
            for="666"
          >
             
            fff
          </label>
        </div>
        <div>
          <input
            id="777"
            type="checkbox"
            value="777"
          />
          <label
            for="777"
          >
             
            ggg
          </label>
        </div>
      </div>
      <div
        class="column"
      >
        <div>
          <input
            id="888"
            type="checkbox"
            value="888"
          />
          <label
            for="888"
          >
             
            hhh
          </label>
        </div>
        <div>
          <input
            id="999"
            type="checkbox"
            value="999"
          />
          <label
            for="999"
          >
             
            iii
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MultiCheck initialize renders the label if label provided 1`] = `
<div>
  <div
    class="MultiCheck"
  >
    <h2>
      my-multi-check
    </h2>
    <div
      class="flex-row"
    >
      <div
        class="column"
      >
        <div>
          <input
            id="select-all"
            type="checkbox"
            value="select-all"
          />
          <label
            for="select-all"
          >
             
            Select All
          </label>
        </div>
        <div>
          <input
            id="111"
            type="checkbox"
            value="111"
          />
          <label
            for="111"
          >
             
            aaa
          </label>
        </div>
        <div>
          <input
            id="222"
            type="checkbox"
            value="222"
          />
          <label
            for="222"
          >
             
            bbb
          </label>
        </div>
        <div>
          <input
            id="333"
            type="checkbox"
            value="333"
          />
          <label
            for="333"
          >
             
            ccc
          </label>
        </div>
        <div>
          <input
            id="444"
            type="checkbox"
            value="444"
          />
          <label
            for="444"
          >
             
            ddd
          </label>
        </div>
        <div>
          <input
            id="555"
            type="checkbox"
            value="555"
          />
          <label
            for="555"
          >
             
            eee
          </label>
        </div>
        <div>
          <input
            id="666"
            type="checkbox"
            value="666"
          />
          <label
            for="666"
          >
             
            fff
          </label>
        </div>
        <div>
          <input
            id="777"
            type="checkbox"
            value="777"
          />
          <label
            for="777"
          >
             
            ggg
          </label>
        </div>
        <div>
          <input
            id="888"
            type="checkbox"
            value="888"
          />
          <label
            for="888"
          >
             
            hhh
          </label>
        </div>
        <div>
          <input
            id="999"
            type="checkbox"
            value="999"
          />
          <label
            for="999"
          >
             
            iii
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;
